module/functions/azure/aad/Assert-AzureServicePrincipalForRbac.ps1

```diff
@@ -1,314 +1,326 @@
+            # Check whether we have an existing credential with the same display name
+            $existingCred = $Application.passwordCredentials |
+                                Where-Object { $_.displayName -eq $CredentialDisplayName }
+        }
+        else {
+            Write-Verbose "Credential will be associated with the Service Principal"
+            $baseUri = "https://graph.microsoft.com/v1.0/servicePrincipals/$($ServicePrincipal.id)"
+            # Check whether we have an existing credential with the same display name
+            $existingCred = $ServicePrincipal.passwordCredentials |
+                                Where-Object { $_.displayName -eq $CredentialDisplayName }
+        }
+
+        # Before adding a credential we need to check if we already added one previously - the number of
+        # credentials for an object can be limited.  We should be a good citizen and remove our old one before
+        # re-generating it
+        if ($existingCred) {
+            Write-Host "Removing existing credential [DisplayName=$CredentialDisplayName; KeyId=$($existingCred.keyId)]"
+            $resp = Invoke-AzRestMethod -Uri "$baseUri/removePassword" `
+                                        -Method POST `
+                                        -Payload ( @{keyId = $existingCred.keyId} | ConvertTo-Json -Compress ) | _HandleRestError
+        }   
+
+        # Now we can generate the new credential - we use the REST API rather than a build cmdlet so that
+        # we can set a display name for the credentials.  This improves traceability and also helps us find
+        # 'our' credential in the future (e.g. when we want to rotate it)
+        $body = @{
+            passwordCredential = @{
+                displayName = $CredentialDisplayName
+                endDateTime = ([DateTime]::Now.AddDays($PasswordLifetimeDays))                    
+            }
+        }
+        $resp = Invoke-AzRestMethod -Uri "$baseUri/addPassword" `
+                                    -Method POST `
+                                    -Payload ($body | ConvertTo-Json -Compress) | _HandleRestError
+        $newCred = $resp.Content |
+                        ConvertFrom-Json -AsHashtable
+        
+
+        # Store the credentials in key vault, if required
+        if ($UseKeyVault) {
+            $appLoginDetails = @{
+                appId = ($applicationMode ? $Application.appId : $ServicePrincipal.appId)
+                password = $newCred.secretText
+                tenant = (Get-AzContext).Tenant.Id
+            }
+            Write-Host "Storing client secret in key vault [VaultName=$KeyVaultName, SecretName=$KeyVaultSecretName]"
+            Set-AzKeyVaultSecret -VaultName $KeyVaultName `
+                                 -Name $KeyVaultSecretName `
+                                 -SecretValue ($appLoginDetails | ConvertTo-Json | ConvertTo-SecureString -AsPlainText -Force) `

```

Comment:
Should the property in `$appLoginDetails` be updated to `clientSecret`? I'm a bit confused as to where the `clientSecret` property is coming from if not from heer?

https://github.com/corvus-dotnet/Corvus.Deployment/pull/109/files#diff-92b916981b3830e7e32d533fcf5335a5b9c0e19872cd130333826c303b93e590R217

Reply:
Yes, I wasn't quite sure how far to go with it.  The original source of this 'chosen' format stems from what used to be required by the `azure/login` GitHub Action - which matched what the Azure CLI produces when you create a service principal via `az ad sp create-for-rbac`.

Now the GH Action [documents](https://github.com/azure/login?tab=readme-ov-file#creds) the usage of `clientSecret` instead, but clearly still supports the old property too (otherwise our usage would be broken in a different way).

The intent for this change was to at least be compatible with both formats, in the event where a secret has been setup manually prior to be serviced via this cmdlet.

I think the options are to either include more the above as a comment in the code, or switch over to using the new property name for any new secrets being created; but I'm not inclined to migrate existing secrets with the old name, but feel free to persuade me otherwise 😀

Reply:
No I don't think migrating the existing secrets is necessary. Maybe just a code comment then with your commentary from above.

Reply:
Having just seen a workflow failure, I see now that the old format does not work with at least the later version of the `azure/login` action we use.  So I've updated the implementation to catch when existing secrets are using the old format and force a rotate, similarly to the previous approach that only looked for the `password` property.